src/domain/core/src/entities/metadata_chain.rs

```diff
@@ -146,6 +147,97 @@ pub trait MetadataChainExt: MetadataChain {
     fn iter_blocks_ref<'a>(&'a self, head: &'a BlockRef) -> DynMetadataStream<'a> {
         self.iter_blocks_interval_ref(head, None)
     }
+
+    async fn accept<E>(
+        &self,
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        self.accept_by_ref(visitors, &BlockRef::Head).await
+    }
+
+    async fn accept_by_hash<E>(
+        &self,
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+        head_hash: &Multihash,
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        let mut decisions = vec![Decision::Next; visitors.len()];
+
+        self.accept_by_hash_with_decisions(&mut decisions, visitors, head_hash)
+            .await
+    }
+
+    async fn accept_by_ref<E>(
+        &self,
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+        head: &BlockRef,
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        let head_hash = self.resolve_ref(head).await.map_err(Into::into)?;
+
+        self.accept_by_hash(visitors, &head_hash).await
+    }
+
+    async fn accept_by_hash_with_decisions<E>(
+        &self,
+        decisions: &mut [Decision],
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = E>],
+        head_hash: &Multihash,
+    ) -> Result<(), E>
+    where
+        E: Error + From<IterBlocksError>,
+    {
+        let mut all_visitors_finished = false;
+        let mut current_hash = Some(head_hash.clone());
+
+        // TODO: PERF: Add traversal optimizations such as skip-lists
+        while let Some(hash) = current_hash
+            && !all_visitors_finished
+        {
+            let block = self.get_block(&hash).await.map_err(Into::into)?;
+            let hashed_block_ref = (&hash, &block);
+
+            let mut stopped_visitors = 0;
+
+            for (decision, visitor) in decisions.iter_mut().zip(visitors.iter_mut()) {
+                match decision {
+                    Decision::Stop => {
+                        stopped_visitors += 1;
+                    }
+                    Decision::NextOfType(type_flags) if type_flags.is_empty() => {
+                        stopped_visitors += 1;
+                    }
+                    Decision::Next => {
+                        *decision = visitor.visit(hashed_block_ref)?;
+                    }
+                    Decision::NextWithHash(requested_hash) => {
+                        if hash == *requested_hash {
+                            *decision = visitor.visit(hashed_block_ref)?;
```

Comment:
So, we are quickly finishing if any of the visitors return an error. I wonder if there would be a need to collect multiple `Result` of visitors. Something like displaying multiple validation errors in 1 run.

Reply:
Moved to https://github.com/kamu-data/kamu-cli/pull/543#discussion_r1517882625